repo-projects/Directory.Build.targets

```diff
@@ -42,23 +45,64 @@
     <WriteLinesToFile File="$(RepoCompletedSemaphorePath)ApplyPatches.complete" Overwrite="true" />
   </Target>
 
+  <Target Name="UpdateNuGetConfig"
+          BeforeTargets="Build"
+          DependsOnTargets="RemoveNuGetConfig;UpdateNuGetConfigWithRepoReferences"
+          Condition="'$(NuGetConfigFile)' != '' OR '@(NuGetConfigFiles)' != ''">
+
+    <WriteLinesToFile File="$(RepoCompletedSemaphorePath)UpdateNuGetConfig.complete" Overwrite="true" />
+  </Target>
+
   <!--
     If a NuGet.config exists in the project, delete it, so the dotnet/source-build nuget.config is
     found instead. This lets the subrepo use the incoming intermediate nupkg contents and avoids the
     need to do subrepo-specific fixups like removing internet sources.
   -->
-  <Target Name="UpdateNuGetConfig"
+  <Target Name="RemoveNuGetConfig"
+          Condition="('$(NuGetConfigFile)' != '' OR '@(NuGetConfigFiles)' != '') and '@(RepositoryReference)' == ''">
+    <ItemGroup>
+      <NuGetConfigFiles Include="$(NuGetConfigFile)" />
+    </ItemGroup>
+
+    <Delete Files="@(NuGetConfigFiles)" />
+  </Target>
+         
+  <Target Name="UpdateNuGetConfigWithRepoReferences"
           BeforeTargets="Build"
-          Condition="'$(NuGetConfigFile)' != '' OR '@(NuGetConfigFiles)' != ''"
+          Condition="('$(NuGetConfigFile)' != '' OR '@(NuGetConfigFiles)' != '') and '@(RepositoryReference)' != ''"
           Inputs="$(MSBuildProjectFullPath)"
           Outputs="$(RepoCompletedSemaphorePath)UpdateNuGetConfig.complete">
+
     <ItemGroup>
       <NuGetConfigFiles Include="$(NuGetConfigFile)" />
     </ItemGroup>
 
-    <Delete Files="@(NuGetConfigFiles)" />
+    <ItemGroup>
+      <_DependentProjectSources Include="@(RepositoryReference -> '%(Identity).proj')">
+        <RepositoryName>%(Identity)</RepositoryName>
+      </_DependentProjectSources>
+    </ItemGroup>
 
-    <WriteLinesToFile File="$(RepoCompletedSemaphorePath)UpdateNuGetConfig.complete" Overwrite="true" />
+    <MSBuild Projects="@(_DependentProjectSources)"
+         Targets="GetPackagesOutput">
+      <Output TaskParameter="TargetOutputs" ItemName="_DependencyProjectsPackagesOutput" />
+    </MSBuild>
+
+    <RemoveInternetSourcesFromNuGetConfig
```

Comment:
This pattern of removing internet sources and adding new sources is not going to handle all situations, for example the there are package source-mappings this would not work.  Did you consider letting the RemoveNuGetConfig target do it's thing and then add a new single root nuget.config with the reference sources?

Reply:
I considered the package source mapping case but decided to punt on it for now since xunit and its submodules don't currently use it.

Reply:
I get that.  My argument is the suggested approach would be simpler, more robust and wouldn't have added a new task to arcade.

Reply:
You mean that we would dynamically create a new NuGet config and add the source to it after the original was deleted?

Reply:
Correct.